### DISCUSSION TITLE
SQE links with full submission ring

### DISCUSSION THREAD

**Discussion by romange** [axboe/liburing#618] [Q&A] [answered]
Suppose I model the operation using SQE links. For example (taken from your latest preso):
`[open file X]→[read file X]→[close file X]`

suppose, I get null when calling `io_uring_get_sqe()` second time (to fill the read request).
Will SQE link still work if I call io_uring_submit() to flush the submit ring together with the first request?

**Comment by axboe** [comment, marked as answer]
It will not, this is also documented in the man pages. Links are only valid within the same submission, they do not persist across different submit calls. If the last request in a submission has IOSQE_IO_LINK (or hard variant) set, then it'll be ignored and still terminate the dependency chain.

**Reply by romange** [reply]
Thanks! May I ask how you then design your linked flows in order to avoid this? By checking the space in the submission buffer in advance?

**Reply by axboe** [reply]
Don't think it's generally been a concern, but I guess it depends on how you manage batches. The SQ ring size must be big enough to accommodate any batch size you want for submit. But yes, you could code it ala:

```
sq_left = io_uring_sq_space_left(ring);

if (sq_left < batch_with_links_I_need_to_submit)
    io_uring_submit(ring);
    
[...]
```

and just flush what you have if you don't have enough SQ entries free for your current linked sequence.

Alternatively, and what I would probably just do, if you're starting a linked sequence then just unconditionally flush what you have already. If you're that close to running out of SQEs, I'd argue you're most likely getting into the territory of holding on to your batches too long and causing higher latencies.
